File: tests/perf/02_csdid_perf.py
"""Track C performance -- Callaway-Sant'Anna staggered DiD.

Times sp.callaway_santanna at N_units in {1000, 5000, 25000} with
T = 5 years. The companion R script times did::att_gt + did::aggte.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import TimingResult, time_repeat, write_results

N_UNITS_LIST = [1_000, 5_000, 25_000]
T = 5
N_REPS = 3  # CS-DiD is heavier; fewer reps


def make_panel(n_units: int, t: int = T, seed: int = 42) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    units = np.repeat(np.arange(n_units), t)
    years = np.tile(np.arange(t), n_units)
    # 60% never-treated, 40% treated split equally across 3 cohorts.
    cohort = np.zeros(n_units, dtype=int)
    treated_idx = rng.choice(n_units, size=int(0.4 * n_units), replace=False)
    rng.shuffle(treated_idx)
    third = len(treated_idx) // 3
    cohort[treated_idx[:third]] = 2
    cohort[treated_idx[third : 2 * third]] = 3
    cohort[treated_idx[2 * third :]] = 4
    first_treat = cohort[units]
    treat = ((first_treat > 0) & (years >= first_treat)).astype(int)
    unit_fe = rng.normal(0, 0.5, size=n_units)
    year_fe = rng.normal(0, 0.3, size=t)
    eps = rng.normal(0, 0.4, size=n_units * t)
    y = unit_fe[units] + year_fe[years] - 0.05 * treat + eps
    return pd.DataFrame(
        {
            "y": y,
            "unit": units,
            "year": years,
            "first_treat": first_treat,
            "treat": treat,
        }
    )


def main() -> None:
    rows: list[TimingResult] = []
    for n_units in N_UNITS_LIST:
        df = make_panel(n_units=n_units)

        def run_sp() -> None:
            sp.callaway_santanna(
                df,
                y="y",
                g="first_treat",
                t="year",
                i="unit",
                estimator="reg",
                control_group="nevertreated",
            )

        med, iqr, mn, mx, peak = time_repeat(run_sp, n_reps=N_REPS, warmup=1)
        rows.append(
            TimingResult(
                estimator="02_csdid",
                side="py",
                n=int(n_units * T),
                n_reps=N_REPS,
                median_time_s=med,
                iqr_time_s=iqr,
                min_time_s=mn,
                max_time_s=mx,
                peak_mem_mb=peak,
                extra={"n_units": n_units, "T": T},
            )
        )
        print(f"  N_units={n_units:>6}  N_obs={n_units*T:>7}  median={med:.3f}s")

    write_results("02_csdid", "py", rows)


if __name__ == "__main__":
    main()

File: tests/perf/02_csdid_perf.R
# Track C performance -- CS-DiD (R side).
# Times did::att_gt + did::aggte at N_units in {1000, 5000, 25000}.

.script_dir <- (function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- grep("^--file=", args, value = TRUE)
  if (length(m) > 0) dirname(normalizePath(sub("^--file=", "", m[1])))
  else getwd()
})()
RESULTS_DIR <- file.path(.script_dir, "results")
dir.create(RESULTS_DIR, showWarnings = FALSE, recursive = TRUE)

suppressPackageStartupMessages({
  library(did)
  library(jsonlite)
})

N_UNITS_LIST <- c(1000L, 5000L, 25000L)
T <- 5L
N_REPS <- 3L

make_panel <- function(n_units, t = T, seed = 42L) {
  set.seed(seed)
  units <- rep(seq_len(n_units), each = t)
  years <- rep(seq_len(t), times = n_units)
  cohort <- integer(n_units)
  treated_idx <- sample.int(n_units, size = as.integer(0.4 * n_units))
  third <- length(treated_idx) %/% 3
  cohort[treated_idx[seq_len(third)]] <- 2L
  cohort[treated_idx[(third + 1L):(2L * third)]] <- 3L
  cohort[treated_idx[(2L * third + 1L):length(treated_idx)]] <- 4L
  first_treat <- cohort[units]
  treat <- as.integer((first_treat > 0L) & (years >= first_treat))
  unit_fe <- rnorm(n_units, 0, 0.5)
  year_fe <- rnorm(t, 0, 0.3)
  eps <- rnorm(n_units * t, 0, 0.4)
  y <- unit_fe[units] + year_fe[years] - 0.05 * treat + eps
  data.frame(y = y, unit = units, year = years,
             first_treat = first_treat, treat = treat)
}

time_one <- function(fn, n_reps, warmup = 1L) {
  for (i in seq_len(warmup)) fn()
  t <- numeric(n_reps)
  for (i in seq_len(n_reps)) {
    gc()
    t0 <- Sys.time(); fn()
    t[i] <- as.numeric(Sys.time() - t0, units = "secs")
  }
  list(median = median(t), iqr = IQR(t), min = min(t), max = max(t))
}

rows <- list()
for (n_units in N_UNITS_LIST) {
  df <- make_panel(n_units)
  fn <- function() {
    fit <- did::att_gt(
      yname = "y", tname = "year", idname = "unit",
      gname = "first_treat", data = df,
      control_group = "nevertreated",
      est_method = "reg", bstrap = FALSE
    )
    did::aggte(fit, type = "simple", bstrap = FALSE, cband = FALSE)
  }
  res <- time_one(fn, n_reps = N_REPS)
  rows[[length(rows) + 1L]] <- list(
    estimator = jsonlite::unbox("02_csdid"),
    side = jsonlite::unbox("R"),
    n = jsonlite::unbox(as.integer(n_units * T)),
    n_reps = jsonlite::unbox(N_REPS),
    median_time_s = jsonlite::unbox(res$median),
    iqr_time_s = jsonlite::unbox(res$iqr),
    min_time_s = jsonlite::unbox(res$min),
    max_time_s = jsonlite::unbox(res$max),
    peak_mem_mb = jsonlite::unbox(NA_real_),
    extra = list(n_units = jsonlite::unbox(n_units), T = jsonlite::unbox(T))
  )
  cat(sprintf("  N_units=%6d  N_obs=%7d  median=%.3fs\n",
              n_units, n_units * T, res$median))
}

payload <- list(
  estimator = jsonlite::unbox("02_csdid"),
  side = jsonlite::unbox("R"),
  rows = rows,
  hardware = list(reference_package = jsonlite::unbox(paste("did", as.character(utils::packageVersion("did")))),
                R_version = jsonlite::unbox(R.version$version.string)),
  extra = list()
)
writeLines(
  jsonlite::toJSON(payload, pretty = TRUE, na = "null", null = "null", digits = NA),
  file.path(RESULTS_DIR, "02_csdid_R.json")
)
message("OK -- wrote 02_csdid_R.json")
